process/project.py

```diff
@@ -4,81 +4,160 @@
-
-            if oldest_child_and_date:
-                main = oldest_child_and_date[0]
-                children.remove(main)
+        self.recordgraph = recordgraph
+
+        # find root entries so we know where to start looking
+        self.roots = defaultdict(list)
+        self.children = defaultdict(list)
+        for entry in entries:
+            node = self.recordgraph.get(entry.fk)
+            if not node:
+                print('Warning: entry not in our record graph: %s from %s'
+                      % (entry.fk, entry.source))
+                continue
+
+            if len(node.parents) == 0:
+                self.roots[entry.source].append(entry)
             else:
-                raise Exception('No main record found for a project %s' % id)
-
-        self.__ppts_main = main
-        self.__ppts_children = children
-
-    @property
-    def main(self):
-        return self.__ppts_main
-
-    @property
-    def children(self):
-        return self.__ppts_children
-
-    def field(self, name):
-        # for ppts, prefer parent record, only moving to children if none
-        # found, at which point we choose the value with the latest
-        # last_updated
-        # TODO: I'm not even sure this is the correct logic to use for dealing
-        # with ambiguities.
-        val = ''
-        if name in self.__ppts_main.values:
-            val = self.__ppts_main.values[name]['value']
-
-        update_date = datetime.min
-        if val == '':
-            for child in self.__ppts_children:
-                if name in child.values:
-                    if child.values[name]['last_updated'] > update_date:
-                        update_date = child.values[name]['last_updated']
-                        val = child.values[name]['value']
-
-        return val
+                self.children[entry.source].append(entry)
+
+        if len(self.roots) == 0:
+            # upgrade oldest child
+            oldest_child = None
+            for (source, entries) in self.children.items():
+                for entry in entries:
+                    if (not oldest_child or
+                            entry.oldest_name_value() <
+                            oldest_child.oldest_name_value()):
+                        oldest_child = entry
+
+            if oldest_child:
+                self.roots[oldest_child.source].append(oldest_child)
+                self.children[oldest_child.source].remove(oldest_child)
+
+    def field(self, name, source):
+        """Fetches the value for a field, using some business logic.
+
+        source is a source.NAME from schemaless.sources
+
+        The process of getting a field:
+          * Start with root project.
+            * If multiple roots, choose one with latest update time.
+          * If source != PPTS, then descend to children.
+            * If multiple children, choose one with latest update time.
+          * If source == PTS, then only descend to children if None found on
```

Comment:
Nit: Not sure you need this if comment, it doesn't seem to really map to the code explicitly?

Reply:
it should. see below (if source != PPTS.NAME)

Reply:
Wouldn't this if for example never be hit (source == PTS) because it would always hit the one above (source != PPTS). The code makes sense to me, it's just the comment that seems odd, might be my reading of it?

Reply:
oh ha, i missed the typo (PTS => PPTS)

Reply:
added in second pull request